### PR TITLE
po/skip already imported

### PR DESCRIPTION
**Pull request by TurboGit**
I don't think reimporting pictures already in the database brings anything when mode is in-place so I have done this unconditionally.

### CODE REVIEW

**Comment by phweyland** [comment]
Deserves some discussion...
I remember having got some issues about already imported images in the past.
If I remember well, a use case is related to xmp, especially when there are multiple (duplicates).
In _image_import_internal() there is a full section about them, which should be removed in that case.

Generally speaking I think this is a good idea. This would simplify / clarify a bit a part of the code.
The user can always save the xmp(s), remove the image and reimport it if necessary.

On the other hand I think it can be a bit misleading not to see the already imported images in the import dialog images list. I would like to see them but disabled (not selectable).

**Comment by elstoc** [comment]
I would prefer this to be optional (with a check box on the import module), and it should certainly work by version number. You still have to check all of the XMPs and only import those whose _version_ does not exist in the database.

One use case is to restore to a previous state from a backup, and it would be nice not to have to remove and re-import.

> I would like to see them but disabled (not selectable).

Or enabled but just not select**ed**.

**Comment by TurboGit** [comment]
I had a look at the code, and in fact we do something when re-importing already present pictures:
- we kill the remove flag from the cache (not sure it is needed)
- we resync the xmp

So I suppose we still want to allow re-importing pictures.

So I suppose we just want to display them in the list but as already present and let the user select it or not. At least it will be easy to quickly see which images is already imported or the new ones. This was the use case I had in mind.

Now we have two options to display them.
1. have a new column with an indicator for new pictures
2. have the new pictures displayed with dimmer color

I prefer solution 1. What do you think ?

**Comment by elstoc** [comment]
> I prefer solution 1. What do you think ?

I agree

**Comment by AlicVB** [comment]
I prefer solution 1 too, as it'll allow to sort by it

**Comment by phweyland** [comment]
+1 for the first one.

**Comment by TurboGit** [comment]
New version pushed which implements solution 1.

**Comment by phweyland** [comment]
For [import in-place] this looks perfect.
But we should not treat/see the flag in [copy and import] not in [import from camera].

**Comment by TurboGit** [comment]
> But we should not treat/see the flag in [copy and import]

That was my first thinking and also what I did before. But actually I think it is interesting to see if some pictures we are about to copy & import are already imported or not in the database.

For the import from camera I fully agree it should not be displayed.

**Comment by TurboGit** [comment]
Well I see also that when you 'copy and import" from a card the column is displayed and this is really confusing. So I'll keep this column only for the in-place indeed.

**Comment by TurboGit** [comment]
Done. Ready for review again. TIA.

**Comment by elstoc** [comment]
If I have an image with multiple versions, I remove one of the versions from the database (leaving the XMP file in place) and re-import the folder, the image shows up as already imported and doesn't re-import unless I select it. 

Personally I'm happy with always re-importing already-imported images (re-reading the XMP files might be unnecessary for most of them but it shouldn't be an issue) and I would prefer that to be the default behaviour (rather than to having to click the right-hand pane and Ctrl+A to select all). Could we make this optional please with a tick box (don't select already-imported files) either at the top of the screen or in the import parameters section of the main module.

**Comment by TurboGit** [comment]
@elstoc : ok for selecting all by defaults. And I'm proposing to add buttons to [unselect all] [select all] [select new] this should help a lot for large lists. Thoughts ?

**Comment by elstoc** [comment]
> And I'm proposing to add buttons to [unselect all] [select all] [select new] this should help a lot for large lists. Thoughts ?

That sounds reasonable. Let's see the implementation though before I comment further :)

**Comment by TurboGit** [comment]
New version proposed with buttons for selection and all selected by default.

**Comment by elstoc** [comment]
I think there's a regression here. If I press escape, I get a toast message saying that images have been imported. Escape should cancel and Enter should import (Enter is managed with a key-press-event handler).

Also, I would suggest, if possible, that the "select all/none/new" buttons should be physically separated from the "import-in-place/cancel" buttons. I think you'd have to build a separate box for the "action area" to do this though.

**Comment by AlicVB** [comment]
hmm... imho, the default action user expect is to *not* reimport already imported images. Reimporting image, can be useful, like @elstoc said, but I think this is a "special" case : repairing some mistake...

maybe we can implement that as a checkbox ? that would have the advantage of being saved from on session to another, so everybody will be happy :)

that said, it's not a showstopper either, even more with @TurboGit new buttons ;)
Annex question : what is the cost in time of reimporting images ?

**Comment by elstoc** [comment]
> maybe we can implement that as a checkbox ?

Yeah I'm fine with that - even hide it away in the collapsible section of the module. I prefer not to have to think too much when I'm importing so I'd rather just always re-import everything, but I can see why others might want it to behave differently.

**Comment by TurboGit** [comment]
@elstoc : Good catch, ESC case fixed.

**Comment by TurboGit** [comment]
> even hide it away in the collapsible section of the module

There is no such section in the in-place mode.

**Comment by TurboGit** [comment]
> Annex question : what is the cost in time of reimporting images ?

Not much I would say, it always seems instantaneous for folders with ~ 200 pictures.

**Comment by elstoc** [comment]
> There is no such section in the in-place mode

I mean in the module on the left-hand-side of the lighttable view

**Comment by TurboGit** [comment]
> I mean in the module on the left-hand-side of the lighttable view, where all of the other preferences are stored.

I see, makes sense indeed. Do you want me to do that or is that new version with buttons fine enough ?

**Comment by elstoc** [comment]
Given that @AlicVB and I disagree on what the default should be, others probably will too. I think a preference would be preferable.

Though I'm happy as long as the default stays as "select all".

**Comment by AlicVB** [comment]
Well, if the cost in time is about to nothing... then maybe don't bother to add that an option... even more with your new button !

**Comment by TurboGit** [comment]
Ok to make everybody happy I have added the preference.

**Comment by elstoc** [comment]
"select new images" doesn't seem very clear to me. You basically have to read the tooltip to know what it means.

I would prefer "select *only* new images" and for the tooltip to say "when using import-in-place, only select images that have not already been imported". No need to say "if activated" - it just makes the sentence harder to parse.

I would still much prefer the 'OK/Cancel' equivalent buttons (Cancel/Import in-place) to be visually separated, either by placing the new buttons above or below the right-hand treeview or with a visual gap in the action bar. This is the norm for most dialogs.

**Comment by phweyland** [comment]
I've made some test. Looks good for me.
The new buttons are handy too.
The only point I disagree with is to put the `select new pictures` flag in module parameters.
This flag has nothing to do with import settings and just controls the behavior of the dialog window, just as `recursive` and `ignore jpegs` do. Then it should be put close to those.

**Comment by phweyland** [comment]
> I would still much prefer the 'OK/Cancel' equivalent buttons (Cancel/Import in-place) to be visually separated, either by placing the new buttons above or below the right-hand treeview or with a visual gap in the action bar. This is the norm for most dialogs.

I agree with the principle. Maybe not easy to implement however.

**Comment by phweyland** [comment]
`select all ` and `select none` could be single toggle button.

**Comment by elstoc** [comment]
> I agree with the principle. Maybe not easy to implement however.

Just do away with the action bar (gtk_dialog_new instead of gtk_dialog_new_with_buttons) and replace it with some buttons that emit the form responses.

**Comment by elstoc** [comment]
It's also worth considering as well that, when you use the built-in action bar, the placement of the buttons/widgets is not necessarily consistent between window managers. Some put the buttons at the bottom, and some put them at the top. If you build them yourself you're guaranteed control over the look/feel.

**Comment by TurboGit** [comment]
> `select all ` and `select none` could be single toggle button.

We don't have a clear semantic as we can have partial selection. So a single button will have to be clicked 2 times get get all or none depending on our implementation.

**Comment by TurboGit** [comment]
> This flag has nothing to do with import settings and just controls the behavior of the dialog window, just as `recursive` and `ignore jpegs` do. Then it should be put close to those.

Agreed, we don't need that in a preset indeed. Will change.

**Comment by TurboGit** [comment]
I also agree that it would be nice to separate the select buttons from the OK/Cancel. I'll try something.

**Comment by TurboGit** [comment]
In latest version, preference moved to dialog. Now the buttons will be difficult to do...

**Comment by elstoc** [comment]
Did you not agree with my tooltip suggestion?

**Comment by TurboGit** [comment]
New button layout proposed.

**Comment by TurboGit** [comment]
> Did you not agree with my tooltip suggestion?

No, just missed that. Now fixed, thanks for the proposal, better indeed.

**Comment by elstoc** [comment]
Now that you've moved the checkbox it doesn't really need the "when using import-in-place" at the start of the tooltip. I think the new buttons probably don't need the "quick select" label (a tooltip might be better) and they could do with a bit of padding between them.

All fairly minor stuff though - seems to work well.

**Comment by AlicVB** [comment]
@TurboGit : feature wise, your last version is great ! really good job !
For the ui details, I find that great too, but @elstoc is certainly more qualified than me :)

**Comment by elstoc** [comment]
> @elstoc is certainly more qualified than me :)

I think the word you're looking for there is picky :)

**Comment by TurboGit** [comment]
@elstoc : remove label for buttons, fix tooltip. if ok I'll merge, let me know.

**Comment by elstoc** [comment]
Nearly! The "select all" and "select none" seem to have some random underscores in the button labels

**Comment by TurboGit** [comment]
@elstoc: stupid copy/paste, thanks for catching, should be fixed now.

**Comment by elstoc** [comment]
Yep all good now

File: src/libs/import.c [hunk 1]
```diff
@@ -1492,10 +1565,51 @@ static void _import_set_collection(char *dirname)
...
+  }
+}
+
 static void _import_from_dialog_run(dt_lib_module_t* self)
 {
   dt_lib_import_t *d = (dt_lib_import_t *)self->data;
-  while(gtk_dialog_run(GTK_DIALOG(d->from.dialog)) == GTK_RESPONSE_ACCEPT)
+  gint res = 0;
+
+  while((res = gtk_dialog_run(GTK_DIALOG(d->from.dialog))) == GTK_RESPONSE_ACCEPT)
```
Comment (phweyland) [inline src/libs/import.c:1612]: the variable res could be removed here

File: src/libs/import.c [hunk 2]
```diff
@@ -1702,7 +1816,8 @@ const struct
   {"session/base_directory_pattern",    "base_pattern",       DT_STRING},
   {"session/sub_directory_pattern",     "sub_pattern",        DT_STRING},
   {"session/filename_pattern",          "filename_pattern",   DT_STRING},
-  {"ui_last/import_initial_rating",     "rating",             DT_INT}
+  {"ui_last/import_initial_rating",     "rating",             DT_INT},
+  {"ui_last/import_select_new",         "select_new",         DT_BOOL}
```
Comment (phweyland) [inline src/libs/import.c:1820]: If `select new` is not in presets, it should be removed from here.

**Review by phweyland — CHANGES_REQUESTED** [review]
small remarks about the code.
The rest seems to be ok for me.
New buttons are in an area which may be problematic with some system layouts as explained by @elstoc, but if necessary we can change that later on.

File: src/libs/import.c (hunk 1 above)
Comment (TurboGit) [reply src/libs/import.c:1612]: Done.

File: src/libs/import.c (hunk 2 above)
Comment (TurboGit) [reply src/libs/import.c:1820]: Done.

**Review by phweyland — APPROVED** [review]

**Comment by phweyland** [comment]
All good, Thanks

**Comment by TurboGit** [comment]
Merged! Thanks for all feedback and review.

**Pull request merged**